### ISSUE TITLE
`zmodem2` crate

### ISSUE THREAD

**Issue by jarkkojs** [hacknus/serial-monitor-rust#90]
Hi just want to race awareness that I've started work on a new ZMODEM crate recently: https://github.com/jarkkojs/zmodem2. It is progressing rapidly as can be witnessed from the Git log. I've also listed issues that I want to reach for my personal uses and purposes for this: https://github.com/jarkkojs/zmodem2/issues.

Would be interesting to see how this could eventually integrated to the serial monitor. Most likely at least https://github.com/jarkkojs/zmodem2/issues/5 and https://github.com/jarkkojs/zmodem2/issues/1 need to be completed before this is doable.

**Comment by hacknus** [comment]
this sounds great, keep me updated!

**Comment by jarkkojs** [comment]
> this sounds great, keep me updated!

I'll ping next time when I get it integrated to my terminal client! So if possible, let's keep this open. I'll try to get a legit usage example in addition to the crate.

**Comment by jarkkojs** [comment]
Almost ready for 0.1 release: https://github.com/jarkkojs/zmodem2/issues

Does not have `async` in the first release but it is per-protocol-frame sequential so you can quite easily implement progress metering etc.

I'll plan to however refactor outside facing API to mimic `TryNext` so that it is easy to glue async later on similarly as I have `std` specific gluing in `std.rs`. There's one issue for this for more elaborate description. You don't want hard bind a crate to async which might be useful e.g. in a boot loader because it is pretty heavy on how much code it generates and all the shenanigans to manage the thread pool but it can be of course supported as a feature later on.

Not rushing with the release just yet though. I'm on hold with one PR for `hex` crate.

Also I might implement cancel protocol to it. You can already of course cancel. That protocol is for notifying the other side of that event.
